### PR TITLE
Ic nd get users endpoint

### PR DESCRIPTION
**Pull request by izzyconner**
ℹ️ Issue #6 

📝 Description 

Primarily, this PR sets up the '/user/users' endpoint for supervisors and admins that allows authenticated users to get a list of users they have access to. **Note that right now, there is no specific 'admin' use that will get all users by default (that's in progress, and will likely come in a later PR)**. For admins, there will need to be a companyId specified in the query. For supervisors, the endpoint will get all user data from their companies by default.

There are currently two ways of getting the user data from cognito: send a separate request to Cognito for each user sub, or get all the users from Cognito and then filter them afterwards by sub. The current setup uses the first method, but there are the beginnings of implementation for the second method in this PR as well. That will be something to shift over to for a later ticket when we are adding in the full implementation for admins.

This PR also sets up a _very_ primitive testing environment setup, where the authentication can be bypassed and a mock user can be used instead to run the endpoints. This needs to be cleaned up in the future, but for now, it works so that we can test the backend directly from Postman (or any other HTTP requesting method of your choice) without having to hook it up to the frontend and manipulate data shown there.

✔️ Testing 

1. Run the backend application alone, making sure to specify a user sub that corresponds to an actual user in Cognito/DynamoDB and `ENV_TYPE=test` in your .env file. The one I was using was `d396491c-22cf-4d63-af1e-4e70e95a29c7`
2. Create a new postman GET request for `http://localhost:3050/user/users` (or, send @izzyconner your postman account and she'll add you to the existing Postman collection for Breaktime). Add any additional queries in that you'd like (companyIds, roles)
3. Once you send the GET request, you should see a list of companyIDs mapped to users returned. 

Testing cases:

- [x] non-admin / non-supervisor request => authentication error
- [x] supervisor request with no queries specified => return list of users for all their companies
- [x] supervisor request with companyId query specified that they belong to => return list of users from that company/companies
- [x]  supervisor request with companyId query specified that they don't belong to => error (e.g. `http://localhost:3050/user/users?companyIds[]=Mcdonalds&companyIds[]=AnotherCompany`)
- [x]  supervisor request with roles for just supervisor => returns just supervisor users for their company/companies
- [x]  supervisor request with roles for associate and supervisor => returns all users for their company/companies (e.g. `http://localhost:3050/user/users?companyIds[]=Mcdonalds&roles[]=associate&roles[]=supervisor`)

🏕️ (Optional) Future Work / Notes 

This PR sets up functionality primarily for supervisors. However, we'll likely need to rework or edit this to include full functionality for admins as well in a separate PR.

We will want to either include throttling for the 30 call limit on Cognito reads or finish and switch to the 'get all users, filter on the backend' way of getting user data from Cognito.

Eventually, we'll want to have more User data returned, likely merging in data from DynamoDB with the Cognito user attributes that are returned. We will also want to distinguish which role/roles they hold for the specific company

We probably want to have a more succinct way of doing the 'testing' environment shift that bypasses the authorization and uses a default mock user - maybe Nest or NodeJS natively has a better way of doing this?

### CODE REVIEW

File: src/aws/cognito/cognito.wrapper.ts [hunk 1]
```diff
@@ -1,20 +1,27 @@
...
+
   verifier = CognitoJwtVerifier.create({
-    userPoolId: process.env.AWS_USER_POOL_ID,  
+    userPoolId: this.userPoolId,  
     tokenUse: "access", 
-    clientId:process.env.AWS_ACCESS_KEY, 
+    clientId: process.env.AWS_ACCESS_KEY, 
   }); 
+
+  // TODO : uhhhhh does this require no credentials? I think that it may have something to with how we set up credentialproviders, but not sure... Can anyone get our users' attributes if they have our region and user pool id?
```
Comment (SofieDunt) [inline src/aws/cognito/cognito.wrapper.ts:23]: @rymaju

File: src/aws/cognito/cognito.wrapper.ts [hunk 2]
```diff
@@ -25,6 +32,69 @@ export class CognitoWrapper {
       console.log("Authentication error! See cognito wrapper"); 
     }
   }
-}
 
+  /**
+   * Gets all user data from the current Cognito user pool.
+   * @returns
```
Comment (SofieDunt) [inline src/aws/cognito/cognito.wrapper.ts:38]: ⛏️ to match below
```suggestion
```

File: src/aws/cognito/cognito.wrapper.ts [hunk 3]
```diff
@@ -25,6 +32,69 @@ export class CognitoWrapper {
...
+      }
+
+      return users;
+    } catch (error) {
+      console.log(error);
+      return [];
+    }
+  }
+
+  async listUsers (userPoolId: string, filter?: string, limit?: number) {
```
Comment (SofieDunt) [inline src/aws/cognito/cognito.wrapper.ts:90]: Can you add doc for this function?

File: src/aws/cognito/cognito.wrapper.ts [hunk 4]
```diff
@@ -25,6 +32,69 @@ export class CognitoWrapper {
...
+
+  async listUsers (userPoolId: string, filter?: string, limit?: number) {  
+    const command = new ListUsersCommand({
+      UserPoolId: userPoolId,
+      Filter: filter,
+      Limit: limit
+    });
+  
+    return this.serviceProvider.send(command)
+      .then((data) => data.Users);
```
Comment (SofieDunt) [inline src/aws/cognito/cognito.wrapper.ts:98]: ❓ Any chance this will result in an error?

File: src/dynamodb.ts [hunk 5]
```diff
@@ -33,9 +36,81 @@ export async function UserTimesheets(uuid:string): Promise<TimeSheetSchema[]> {
...
+  // There should only ever be one entry per user in the table (since there's no sort key, the partition key is the only unique index used)
+  if (userCompaniesData.length != 1) {
+    throw new Error('Invalid entries in DynamoDB, should only have a single entry per user');
+  }
+
+  return userCompaniesData[0];
+}
+
+/**
+ *
```
Comment (SofieDunt) [inline src/dynamodb.ts:81]: ❓Fill in?

File: src/users/users.controller.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  
+  // TODO : For now, this primarily just works for supervisors, rather than admins. Some additional checks will need to be added for full admin support
+  /**
+   * Gets all the user data from a certain company/companies and returns and array. If no company ID is given, 
+   * we'll default to use the companies that the requesting user belongs to as the filter. If the user is an admin, return all users
+   * unless a company ID is specified.
+   * 
+   * @param companyID an optional filter query to get users only from this company, comma-delineated
+   * @param roles an filter query to specify which users to get by roles (associate, supervisor, and/or admin); by default, will be only associates
+   * @throws some error if the companyID requested is not one that the user belongs, maybe 403 forbidden??
```
Comment (SofieDunt) [inline src/users/users.controller.ts:22]: Let's do 401 unauthorized

File: thoughts.txt [hunk 7]
```diff
@@ -0,0 +1,79 @@
+Options:
```
Comment (SofieDunt) [inline thoughts.txt:1]: 💭 Maybe we should move this to docs or a GH pages

**Review by SofieDunt — COMMENTED** [review]
Code looks great!

**Review by circlegov — APPROVED** [review]
lgtm

File: thoughts.txt (hunk 7 above)
Comment (izzyconner) [reply thoughts.txt:1]: whoops yep didn't realize I committed our notes lol

**Review by circlegov — APPROVED** [review]
lgtm

**Review by SofieDunt — APPROVED** [review]
documentation 😍 
💯 Great work on this!

**Pull request merged**